lib/snapshot/runner.rb

```diff
@@ -55,6 +51,24 @@ def work
       FileUtils.rm_rf(TestCommandGenerator.derived_data_path)
     end
 
+    # This is its own method so that it can re-try if the tests fail randomly
+    # @return true/false depending on if the tests succeded
+    def run_for_device_and_language(language, device, launch_arguments, retries = 0)
+      return launch(language, device, launch_arguments)
+    rescue => ex
+      UI.error ex.to_s # we should to show right here as well
+
+      if retries < Snapshot.config[:number_of_retries]
+        UI.important "Tests failed, re-trying #{retries + 1} out of #{Snapshot.config[:number_of_retries] + 1} times"
+        run_for_device_and_language(language, device, launch_arguments, retries + 1)
+      else
+        UI.error "Backtrace:\n\t#{ex.backtrace.join("\n\t")}" if $verbose
+        self.collected_errors << ex
+        raise ex if Snapshot.config[:stop_after_first_error]
```

Comment:
These two commandline options now seem a bit confusing.  I'm not sure how to solve it.  This option would probably be better as `stop_after_first_complete_failure` and the new one `number_of_retries_for_failed_test` but changing the existing option would not be backward compatible.